drivers/sdhc/sam_sdmmc.c

```diff
@@ -0,0 +1,1091 @@
+#include <soc.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sdmmc, CONFIG_SDHC_LOG_LEVEL);
+
+#define CMD_ERROR_MASK	(SDMMC_EISTR_SD_SDIO_CMDIDX_Msk | \
+			 SDMMC_EISTR_SD_SDIO_CMDEND_Msk | \
+			 SDMMC_EISTR_SD_SDIO_CMDCRC_Msk | \
+			 SDMMC_EISTR_SD_SDIO_CMDTEO_Msk)
+#define DATA_ERROR_MASK	(SDMMC_EISTR_SD_SDIO_DATEND_Msk | \
+			 SDMMC_EISTR_SD_SDIO_DATCRC_Msk | \
+			 SDMMC_EISTR_SD_SDIO_DATTEO_Msk)
+
+#define INT_CMD_MASK	(SDMMC_NISTR_SD_SDIO_TRFC_Msk | \
+			 SDMMC_NISTR_SD_SDIO_CMDC_Msk)
+#define INT_DATA_MASK	(SDMMC_NISTR_SD_SDIO_BRDRDY_Msk | \
+			 SDMMC_NISTR_SD_SDIO_BWRRDY_Msk | \
+			 SDMMC_NISTR_SD_SDIO_DMAINT_Msk)
+#define INT_CMD_ERROR_MASK \
+			(CMD_ERROR_MASK | \
+			 SDMMC_EISTR_SD_SDIO_ACMD_Msk)
+#define INT_DATA_ERROR_MASK \
+			(DATA_ERROR_MASK | \
+			 SDMMC_EISTR_SD_SDIO_ADMA_Msk)
+
+#define INT_MASK		(INT_CMD_MASK | \
+				 INT_DATA_MASK | \
+				 SDMMC_NISTR_SD_SDIO_ERRINT_Msk | \
+				 SDMMC_NISTR_SD_SDIO_CREM_Msk | \
+				 SDMMC_NISTR_SD_SDIO_CINS_Msk)
+#define INT_ERROR_MASK		(INT_CMD_ERROR_MASK | \
+				 INT_DATA_ERROR_MASK | \
+				 SDMMC_EISTR_SD_SDIO_CURLIM_Msk)
+
+#define DEFAULT_INT_MASK	(INT_CMD_MASK | \
+				 SDMMC_NISTR_SD_SDIO_CINS_Msk | \
+				 SDMMC_NISTR_SD_SDIO_CREM_Msk)
+#define DEFAULT_INT_ERROR_MASK	(CMD_ERROR_MASK | \
+				 DATA_ERROR_MASK | \
+				 SDMMC_EISTR_SD_SDIO_CURLIM_Msk)
+
+#define IS_READ_CMD(cmd) \
+((((cmd) == SD_READ_SINGLE_BLOCK) || ((cmd) == SD_READ_MULTIPLE_BLOCK)) ? 1 : 0)
+#define IS_WRITE_CMD(cmd) \
+((((cmd) == SD_WRITE_SINGLE_BLOCK) || ((cmd) == SD_WRITE_MULTIPLE_BLOCK)) ? 1 : 0)
+#define IS_MULTI_BLOCK(cmd) \
+((((cmd) == SD_READ_MULTIPLE_BLOCK) || ((cmd) == SD_WRITE_MULTIPLE_BLOCK)) ? 1 : 0)
+#define IS_DATA_CMD(cmd) \
+((IS_READ_CMD(cmd) || IS_WRITE_CMD(cmd)) ? 1 : 0)
+
+#define INH_TO_RST(x)	(((x) & 0x3) << 1)
+
+/* ADMA2 32-bit descriptor */
+struct adma2_desc {
+	uint16_t cmd;
+	uint16_t len;
+	uint32_t addr;
+} __packed __aligned(4);
+
+#define ADMA2_TRAN_VALID	0x21
+#define ADMA2_INT		0x4
+#define ADMA2_END		0x2
+
+/* ADMA2 data alignment */
+#define ADMA2_ALIGN		4
+#define ADMA2_MASK		(ADMA2_ALIGN - 1)
+#define ADMA2_ALIGNED(addr)	(!(((uint32_t)(addr)) & ADMA2_MASK))
+
+#define ADMA2_NUM_DESC	64
+#define ADMA2_MAX_LEN	65536
+#define ADMA2_MAX_SIZE	(ADMA2_NUM_DESC * ADMA2_MAX_LEN)
+
+enum req_flag {
+	REQ_RSP_PRESENT	= BIT(0),
+	REQ_RSP_136	= BIT(1),
+	REQ_RSP_DATA	= BIT(2),
+	REQ_RSP_TRFC	= BIT(3),
+	REQ_USE_ADMA	= BIT(4),
+	REQ_IS_READ	= BIT(5),
+};
+
+#define REQ_RSP_MASK	(REQ_RSP_PRESENT | \
+			 REQ_RSP_136 | \
+			 REQ_RSP_DATA | \
+			 REQ_RSP_TRFC)
+
+struct sd_request {
+	enum req_flag flags;
+	uint32_t reset_mask;
+	uint32_t *response;
+	uint32_t block_size;
+	uint32_t blocks;
+	uint8_t  *data;
+	struct   k_sem completion;
+};
+
+struct sam_sdmmc_config {
+	sdmmc_registers_t *base;
+	const struct atmel_sam_pmc_config clock_cfg;
+	const struct pinctrl_dev_config *pincfg;
+	uint32_t base_clk;
+	uint8_t  non_removable;
```

Comment:
```suggestion
	unsigned int  non_removable : 1;
```
nit- these fields and others could probably be stored as bitfield flags to save some space